src/main/scala/TypeCheck.scala

```diff
@@ -0,0 +1,138 @@
+package fuselang
+
+import Syntax._
+import Errors._
+
+object TypeChecker {
+  import TypeEnv._
+
+  def typeCheck(p: Prog) = {
+    val initEnv = p.decls.foldLeft(emptyEnv)({ case (env, Decl(id, typ)) =>
+      id.typ = Some(typ);
+      env.addBind(id -> Info(id, typ))
+    })
+    checkC(p.cmd)(initEnv, List[Id]())
+  }
+
+  private def checkB(t1: Type, t2: Type, op: Op2) = op match {
+    case OpEq() | OpNeq() => {
+      if (t1 :< t2 || t2 :< t1) TBool()
+      else throw UnexpectedSubtype(op.pos, op.toString, t1, t2)
+    }
+    case OpLt() | OpLte() | OpGt() | OpGte() => (t1, t2) match {
+      case ((TStaticInt(_) | TSizedInt(_)), (TStaticInt(_) | TSizedInt(_))) => TBool()
+      case (_: TFloat, _: TFloat) => TFloat()
+      case _ => throw BinopError(op, t1, t2)
+    }
+    case OpAdd() | OpTimes() | OpSub() | OpDiv() => (t1, t2) match {
+      case ((TStaticInt(_) | TSizedInt(_)), (TStaticInt(_) | TSizedInt(_))) => t1.join(t2, op.toFun)
+      case (_: TFloat, _: TFloat) => TFloat()
+      case _ => throw BinopError(op, t1, t2)
+    }
+
+  }
+
+  private def checkE(expr: Expr)(implicit env: Env, its: ItStack): (Type, Env) = expr match {
+    case EFloat(_) => TFloat() -> env
+    case EInt(v) => TStaticInt(v) -> env
+    case EBool(_) => TBool() -> env
+    case EVar(id) => {
+      // Add type information to variable
+      id.typ = Some(env(id).typ);
+      env(id).typ -> env
+    }
+    case EBinop(op, e1, e2) => {
+      val (t1, env1) = checkE(e1)
+      val (t2, env2) = checkE(e2)(env1, its)
+      checkB(t1, t2, op) -> env2
+    }
+    case EAA(id, idxs) => env(id).typ match {
+      case TArray(typ, dims) => {
+        if (dims.length != idxs.length) {
+          throw IncorrectAccessDims(id, dims.length, idxs.length)
+        }
+        // update type for id
+        id.typ = Some(env(id).typ);
```

Comment:
As above—let's be careful about mutating the AST while checking it.

Reply:
Added comment to type checker about mutating behavior.